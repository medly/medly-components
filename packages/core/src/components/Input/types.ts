import { HTMLProps } from '@medly-components/utils';
import { LabelPositions } from '../Label/types';

export interface InputProps extends HTMLProps<HTMLInputElement> {
    /** Label Position */
    labelPosition?: LabelPositions;
    /** Takes full width of the parent component */
    fullWidth?: boolean;
    /** Prefix  */
    prefix?: string;
    /** Suffix */
    suffix?: string;
    /** Disable Input */
    disabled?: boolean;
    /** To be used when it is required in any form */
    required?: boolean;
    /** Input label */
    label?: string;
    /** Input description */
    description?: string;
    /** Description color */
    descriptionColor?: string;
}
