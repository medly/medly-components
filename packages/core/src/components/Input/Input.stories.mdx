import { Input } from './Input';
import { Preview, Story, Meta, Props } from '@storybook/addon-docs/blocks';
import * as stories from './Input.stories';

<Meta title="Core" component={Input} />

# Input

The `Input` component allows you to get user input in a text field. 

> **Note:**  Input is deprecated in favor of the TextField component.

<Preview withToolbar>
    <Story
        name="Input"
        parameters={{
            status: {
                type: 'deprecated'
            }
        }}
    >
        {stories.Basic()}
    </Story>
</Preview>

## Props

This component offers the following props mentioned in the table below:

<Props of={Input} />

## InputTheme

This component offers the following themes mentioned in the table below:

<Props of={stories.ThemeInterface} />
