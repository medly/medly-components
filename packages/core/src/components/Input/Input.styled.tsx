import { fullWidth, getFontStyle, styled } from '@medly-components/utils';
import FieldWithLabel from '../FieldWithLabel';
import { InputProps } from './types';

export const Wrapper = styled(FieldWithLabel.Field).attrs(({ theme: { input, font } }) => ({ input, font }))<{
    fullWidth?: boolean;
}>`
    display: flex;
    &&& {
        flex-direction: row;
        align-items: center;
    }
    background-clip: padding-box;
    box-sizing: border-box;
    border: 1px solid ${props => props.input.borderColor};
    border-radius: 5px;
    overflow: hidden;
    width: ${props => (props.fullWidth ? '100%' : 'max-content')};
    height: max-content;

    &:focus-within {
        border-color: ${props => props.input.focusedBorderColor};
    }

    & > * {
        ${getFontStyle}
        padding: ${({ theme }) => theme.spacing.S2};
        color: ${props => props.input.color};
    }
`;

export const Input = styled('input').attrs(({ theme: { input, font } }) => ({ input, font }))<InputProps>`
    flex: 1;
    outline: 0;
    border: 0;
    ${fullWidth('16px')};
    cursor: ${({ disabled }) => (disabled ? 'not-allowed' : 'inherit')};
    background-color: ${({ disabled, input }) => (disabled ? input.disabledBgcolor : input.bgColor)};

    &::placeholder {
        color: ${props => props.input.placeHolderColor};
    }

    &::-webkit-file-upload-button {
        visibility: hidden;
    }
`;

export const Prefix = styled('span')`
    background-color: ${({ theme }) => theme.input.prefixAndSuffixBgColor};
`;

export const Suffix = styled('span')`
    background-color: ${({ theme }) => theme.input.prefixAndSuffixBgColor};
`;
