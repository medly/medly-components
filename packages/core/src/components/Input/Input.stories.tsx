import { defaultTheme, InputTheme } from '@medly-components/theme';
import { boolean, color, select, text } from '@storybook/addon-knobs';
import React from 'react';
import { Input } from './Input';
import { InputProps } from './types';

const labelPositions: Required<InputProps>['labelPosition'][] = ['top', 'bottom', 'left', 'right'];

export const ThemeInterface: React.FC<InputTheme> = () => null;
ThemeInterface.defaultProps = {
    ...defaultTheme.input
};

export const Basic = () => (
    <Input
        id="storybook-input"
        type="email"
        fullWidth={boolean('Full Width', false)}
        disabled={boolean('Disabled', false)}
        label={text('Label', 'Email Address')}
        labelPosition={select('Label Position', labelPositions, 'top')}
        required={boolean('Required', false)}
        placeholder={text('Placeholder', 'Enter Email Address')}
        description={text('Description', 'We will never share your email with anyone')}
        prefix={text('Prefix', '')}
        suffix={text('Suffix', '.com')}
        descriptionColor={color('Description Color', defaultTheme.input.descriptionColor)}
    />
);
