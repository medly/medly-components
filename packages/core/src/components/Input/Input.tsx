import { WithStyle } from '@medly-components/utils';
import React, { FC } from 'react';
import FieldWithLabel from '../FieldWithLabel';
import * as Styled from './Input.styled';
import { InputProps } from './types';

const Component: FC<InputProps> = React.memo(
    React.forwardRef((props, ref) => {
        const { description, label, labelPosition, fullWidth, required, descriptionColor, prefix, suffix, ...restProps } = props,
            id = props.id || 'medly-input';
        return (
            <FieldWithLabel id={`${id}-field`} {...{ fullWidth, labelPosition }}>
                {label && (
                    <FieldWithLabel.Label htmlFor={id} {...{ required, labelPosition }}>
                        {label}
                    </FieldWithLabel.Label>
                )}
                <Styled.Wrapper fullWidth={fullWidth}>
                    {prefix && <Styled.Prefix>{prefix}</Styled.Prefix>}
                    <Styled.Input id={id} ref={ref} required={required} fullWidth={fullWidth} {...restProps} />
                    {suffix && <Styled.Suffix>{suffix}</Styled.Suffix>}
                </Styled.Wrapper>
                {description && <FieldWithLabel.Description textColor={descriptionColor}>{description}</FieldWithLabel.Description>}
            </FieldWithLabel>
        );
    })
);

Component.displayName = 'Input';
Component.defaultProps = {
    type: 'text',
    disabled: false,
    fullWidth: false,
    required: false,
    label: '',
    labelPosition: 'top',
    description: ''
};
export const Input: FC<InputProps> & WithStyle = Object.assign(Component, { Style: Styled.Input });
