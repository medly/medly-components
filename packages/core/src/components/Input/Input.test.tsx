import { render } from '@test-utils';
import { Input } from './Input';

describe('Input component', () => {
    it('should render correctly with all the props given', () => {
        const { container } = render(
            <Input
                id="email-input"
                type="email"
                fullWidth
                label="Email Address"
                labelPosition="top"
                required
                placeholder="Enter Email Address"
                prefix="www."
                suffix=".com"
                description="We will never share your email with anyone"
                descriptionColor="grey"
            />
        );
        expect(container).toMatchSnapshot();
    });

    it('should render correctly when fullWidth & required props are falsy', () => {
        const { container } = render(
            <Input
                type="email"
                label="Email Address"
                labelPosition="left"
                placeholder="Enter Email Address"
                description="We will never share your email with anyone"
            />
        );
        expect(container).toMatchSnapshot();
    });

    it('should render correctly when label and description are not given', () => {
        const { container } = render(<Input id="email-input" type="email" placeholder="Enter Email Address" />);
        expect(container).toMatchSnapshot();
    });

    it('should render correctly when it is disabled', () => {
        const { container } = render(<Input id="email-input" disabled placeholder="Enter Email Address" />);
        expect(container).toMatchSnapshot();
    });
});
