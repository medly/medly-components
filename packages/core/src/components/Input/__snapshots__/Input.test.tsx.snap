// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input component should render correctly when fullWidth & required props are falsy 1`] = `
.c1 {
  color: #13181D;
  font-size: 1.4rem;
  font-weight: 400;
  -webkit-letter-spacing: 0rem;
  -moz-letter-spacing: 0rem;
  -ms-letter-spacing: 0rem;
  letter-spacing: 0rem;
  line-height: 2.2rem;
  margin: 0px 0.8rem 0px 0px;
  cursor: default;
}

.c5 {
  margin: 0;
  color: inherit;
  font-size: 1.4rem;
  font-weight: 400;
  -webkit-letter-spacing: 0rem;
  -moz-letter-spacing: 0rem;
  -ms-letter-spacing: 0rem;
  letter-spacing: 0rem;
  line-height: 2.2rem;
  text-align: initial;
  display: inline-block;
}

.c2 {
  grid-area: label;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c6 {
  grid-area: description;
  margin: 5px 0 0 2px;
  color: #012040;
}

.c0 {
  margin: 0.8rem 0.8rem 0.8rem 0;
  display: inline-grid;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  grid-template-columns: minmax(max-content,25%) minmax(auto,80%);
  grid-template-areas: 'label field' 'random description';
}

.c0 .sc-12atfcu-0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.c3 {
  grid-area: field;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  margin-left: 0px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  background-clip: padding-box;
  box-sizing: border-box;
  border: 1px solid #b0bcc8;
  border-radius: 5px;
  overflow: hidden;
  width: -webkit-max-content;
  width: -moz-max-content;
  width: max-content;
  height: -webkit-max-content;
  height: -moz-max-content;
  height: max-content;
}

.c3.c3.c3 {
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c3:focus-within {
  border-color: #012040;
}

.c3 > * {
  font-size: 1.4rem;
  font-weight: 400;
  -webkit-letter-spacing: 0rem;
  -moz-letter-spacing: 0rem;
  -ms-letter-spacing: 0rem;
  letter-spacing: 0rem;
  line-height: 2.2rem;
  font-family: Open Sans,Helvetica Neue,Helvetica,Arial,sans-serif;
  padding: 0.8rem;
  color: #435465;
}

.c4 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  outline: 0;
  border: 0;
  width: calc(100% - 16px);
  cursor: inherit;
  background-color: #ffffff;
}

.c4::-webkit-input-placeholder {
  color: #98A7B7;
}

.c4::-moz-placeholder {
  color: #98A7B7;
}

.c4:-ms-input-placeholder {
  color: #98A7B7;
}

.c4::placeholder {
  color: #98A7B7;
}

.c4::-webkit-file-upload-button {
  visibility: hidden;
}

<div>
  <div
    class="c0"
    id="medly-input-field"
  >
    <label
      class="c1 c2"
      for="medly-input"
    >
      Email Address
    </label>
    <div
      class="sc-12atfcu-0 c3"
    >
      <input
        class="c4"
        id="medly-input"
        placeholder="Enter Email Address"
        type="email"
      />
    </div>
    <span
      class="c5 c6"
    >
      We will never share your email with anyone
    </span>
  </div>
</div>
`;

exports[`Input component should render correctly when it is disabled 1`] = `
.c0 {
  margin: 0.8rem 0.8rem 0.8rem 0;
  display: inline-grid;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  grid-template-areas: 'label' 'field' 'description';
  grid-template-columns: auto;
  grid-template-areas: 'field' 'description';
}

.c0 .sc-12atfcu-0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c1 {
  grid-area: field;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  margin-left: 0px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  background-clip: padding-box;
  box-sizing: border-box;
  border: 1px solid #b0bcc8;
  border-radius: 5px;
  overflow: hidden;
  width: -webkit-max-content;
  width: -moz-max-content;
  width: max-content;
  height: -webkit-max-content;
  height: -moz-max-content;
  height: max-content;
}

.c1.c1.c1 {
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1:focus-within {
  border-color: #012040;
}

.c1 > * {
  font-size: 1.4rem;
  font-weight: 400;
  -webkit-letter-spacing: 0rem;
  -moz-letter-spacing: 0rem;
  -ms-letter-spacing: 0rem;
  letter-spacing: 0rem;
  line-height: 2.2rem;
  font-family: Open Sans,Helvetica Neue,Helvetica,Arial,sans-serif;
  padding: 0.8rem;
  color: #435465;
}

.c2 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  outline: 0;
  border: 0;
  width: calc(100% - 16px);
  cursor: not-allowed;
  background-color: #C7D0D8;
}

.c2::-webkit-input-placeholder {
  color: #98A7B7;
}

.c2::-moz-placeholder {
  color: #98A7B7;
}

.c2:-ms-input-placeholder {
  color: #98A7B7;
}

.c2::placeholder {
  color: #98A7B7;
}

.c2::-webkit-file-upload-button {
  visibility: hidden;
}

<div>
  <div
    class="c0"
    id="email-input-field"
  >
    
    <div
      class="sc-12atfcu-0 c1"
    >
      <input
        class="c2"
        disabled=""
        id="email-input"
        placeholder="Enter Email Address"
        type="text"
      />
    </div>
    
  </div>
</div>
`;

exports[`Input component should render correctly when label and description are not given 1`] = `
.c0 {
  margin: 0.8rem 0.8rem 0.8rem 0;
  display: inline-grid;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  grid-template-areas: 'label' 'field' 'description';
  grid-template-columns: auto;
  grid-template-areas: 'field' 'description';
}

.c0 .sc-12atfcu-0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c1 {
  grid-area: field;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  margin-left: 0px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  background-clip: padding-box;
  box-sizing: border-box;
  border: 1px solid #b0bcc8;
  border-radius: 5px;
  overflow: hidden;
  width: -webkit-max-content;
  width: -moz-max-content;
  width: max-content;
  height: -webkit-max-content;
  height: -moz-max-content;
  height: max-content;
}

.c1.c1.c1 {
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1:focus-within {
  border-color: #012040;
}

.c1 > * {
  font-size: 1.4rem;
  font-weight: 400;
  -webkit-letter-spacing: 0rem;
  -moz-letter-spacing: 0rem;
  -ms-letter-spacing: 0rem;
  letter-spacing: 0rem;
  line-height: 2.2rem;
  font-family: Open Sans,Helvetica Neue,Helvetica,Arial,sans-serif;
  padding: 0.8rem;
  color: #435465;
}

.c2 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  outline: 0;
  border: 0;
  width: calc(100% - 16px);
  cursor: inherit;
  background-color: #ffffff;
}

.c2::-webkit-input-placeholder {
  color: #98A7B7;
}

.c2::-moz-placeholder {
  color: #98A7B7;
}

.c2:-ms-input-placeholder {
  color: #98A7B7;
}

.c2::placeholder {
  color: #98A7B7;
}

.c2::-webkit-file-upload-button {
  visibility: hidden;
}

<div>
  <div
    class="c0"
    id="email-input-field"
  >
    
    <div
      class="sc-12atfcu-0 c1"
    >
      <input
        class="c2"
        id="email-input"
        placeholder="Enter Email Address"
        type="email"
      />
    </div>
    
  </div>
</div>
`;

exports[`Input component should render correctly with all the props given 1`] = `
.c1 {
  color: #13181D;
  font-size: 1.4rem;
  font-weight: 400;
  -webkit-letter-spacing: 0rem;
  -moz-letter-spacing: 0rem;
  -ms-letter-spacing: 0rem;
  letter-spacing: 0rem;
  line-height: 2.2rem;
  margin: 0px 0px 0.8rem 0px;
  cursor: default;
}

.c1::after {
  color: red;
  content: ' *';
}

.c7 {
  margin: 0;
  color: grey;
  font-size: 1.4rem;
  font-weight: 400;
  -webkit-letter-spacing: 0rem;
  -moz-letter-spacing: 0rem;
  -ms-letter-spacing: 0rem;
  letter-spacing: 0rem;
  line-height: 2.2rem;
  text-align: initial;
  display: inline-block;
}

.c2 {
  grid-area: label;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c8 {
  grid-area: description;
  margin: 5px 0 0 2px;
  color: grey;
}

.c0 {
  margin: 0.8rem 0;
  display: grid;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  grid-template-areas: 'label' 'field' 'description';
}

.c0 .sc-12atfcu-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c3 {
  grid-area: field;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  margin-left: 0px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  background-clip: padding-box;
  box-sizing: border-box;
  border: 1px solid #b0bcc8;
  border-radius: 5px;
  overflow: hidden;
  width: 100%;
  height: -webkit-max-content;
  height: -moz-max-content;
  height: max-content;
}

.c3.c3.c3 {
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c3:focus-within {
  border-color: #012040;
}

.c3 > * {
  font-size: 1.4rem;
  font-weight: 400;
  -webkit-letter-spacing: 0rem;
  -moz-letter-spacing: 0rem;
  -ms-letter-spacing: 0rem;
  letter-spacing: 0rem;
  line-height: 2.2rem;
  font-family: Open Sans,Helvetica Neue,Helvetica,Arial,sans-serif;
  padding: 0.8rem;
  color: #435465;
}

.c5 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  outline: 0;
  border: 0;
  width: calc(100% - 16px);
  cursor: inherit;
  background-color: #ffffff;
}

.c5::-webkit-input-placeholder {
  color: #98A7B7;
}

.c5::-moz-placeholder {
  color: #98A7B7;
}

.c5:-ms-input-placeholder {
  color: #98A7B7;
}

.c5::placeholder {
  color: #98A7B7;
}

.c5::-webkit-file-upload-button {
  visibility: hidden;
}

.c4 {
  background-color: #eff2f4;
}

.c6 {
  background-color: #eff2f4;
}

<div>
  <div
    class="c0"
    id="email-input-field"
  >
    <label
      class="c1 c2"
      for="email-input"
      required=""
    >
      Email Address
    </label>
    <div
      class="sc-12atfcu-0 c3"
    >
      <span
        class="c4"
      >
        www.
      </span>
      <input
        class="c5"
        id="email-input"
        placeholder="Enter Email Address"
        required=""
        type="email"
      />
      <span
        class="c6"
      >
        .com
      </span>
    </div>
    <span
      class="c7 c8"
    >
      We will never share your email with anyone
    </span>
  </div>
</div>
`;
